File: Bit_Manipulation/python/insertion.py
from numpy import uint32


def get_bit(num, i):
    return ((num & (1 << i)) != 0)


def set_bit(num, i):
    return num | (1 << i)


def clear_bit(num, i):
    mask = ~(1 << i)

    return num & mask


def clear_middle(num, i, j):
    mask = 0
    for shift in range(j - i):
        mask += (1 << shift + i)
    return num & ~mask


def insert(n, m, i, j):
    n = clear_middle(n, i, j)

    for shift in range(j - i + 1):
        if get_bit(m, shift):
            n = set_bit(n, shift + i)

    return n

def main():
    num1 = uint32(1024)
    num2 = uint32(19)
    print("{0:b}".format(num1))
    print("{0:b}".format(num2))

    result = insert(num1, num2, 2, 6)
    print("{0:b}".format(result))


if __name__ == '__main__':
    main()

File: Bit_Manipulation/go/insertion.go
package main

import (
  "fmt"
)


func setBit(num *uint32, i int) {
  *num = *num | (1 << i)
}


func getBit(num uint32, i int) bool {
  return ((num & (1 << i)) != 0)
}


func clearMiddle(num *uint32, i int, j int) {
  var mask uint32

  for shift := 0; shift < (j - i); shift++ {
    mask = mask + (1 << uint32(shift + i))
  }

  *num = *num & ^mask
}


func insert(n *uint32, m uint32, i int, j int) {
  clearMiddle(n, i, j)

  for shift := 0; shift < (j - i + 1); shift++ {
    if getBit(m, shift) {
      setBit(n, shift + i)
    }
  }
}


func main() {
  var a, b uint32
  a = 1024
  b = 18

  fmt.Printf("%032b\n", a)
  fmt.Printf("%032b\n\n", b)

  insert(&a, b, 2, 6)
  fmt.Printf("%032b\n", a)
}
